File: ios/MemberApp/Features/Scan/CheckinStore.swift
import Foundation
import Observation

/// The organiser side of an event: who registered, and what they answered.
///
/// Everything here is read with the *staff member's own* Indico authorization —
/// there is no shared credential and no server of ours in the middle. Indico
/// decides who may see it: the check-in API is `RHManageEventBase`, so a member
/// without management rights on the event gets a 403 and never sees the door.
/// That 403 is the permission model; the app only asks.
///
/// Reading needs only `read:everything`. Recording a check-in is a `PATCH`, and
/// Indico accepts only `registrants` or `full:everything` for anything that is
/// not a GET — so [checkIn] needs the wider grant, and asks for it on the door
/// screen alone. Members are never re-authorized for it: Indico extends an
/// existing authorization in place, so only the staffer who opens the door
/// consents, and only once. See `IndicoAuthConfiguration.checkinScopes`.
@MainActor
@Observable
final class CheckinStore {
    enum Access: Equatable {
        case unknown
        case checking
        /// This member manages the event and may read its registrations.
        case allowed
        case denied
    }

    /// A registrant, plus the form they belong to — the detail lookup needs both.
    struct Entry: Equatable, Identifiable {
        let formID: Int
        let registration: CheckinRegistration
        var id: Int { registration.id }
    }

    /// One registration form on an event.
    ///
    /// An event can carry several, and Indico keeps them genuinely apart: a
    /// registration lives inside exactly one form, and `checked_in` is a column
    /// on the registration. 烤場集合 runs a 報名表 and a 遊覽車報名表, so a member
    /// who booked the coach as well holds two registrations, two tickets, and
    /// two check-ins to be made at two different desks.
    ///
    /// Which is why a door is opened per form rather than per event. Pouring
    /// them into one list gave a roster where the same person appeared twice, a
    /// denominator no door could ever reach, and a scan that recorded
    /// attendance against whichever of the two Indico happened to return first.
    struct Form: Identifiable, Equatable {
        let id: Int
        let title: String
        /// Whether Indico is still taking registrations on it.
        let isOpen: Bool
        /// Indico's own numbers, as of the probe.
        ///
        /// `registration_count` is `existing_registrations_count`: it counts the
        /// withdrawn and rejected rows that [expected] drops, and counts an
        /// accompanying person as a seat. So it is the right number to show
        /// before the roster lands and the wrong one to keep once it has, which
        /// is why the count helpers prefer the roster and fall back to this.
        let registrationCount: Int
        let checkedInCount: Int
    }

    private(set) var access: [String: Access] = [:]
    private(set) var roster: [String: [Entry]] = [:]
    /// The event's forms, learned from the same probe that answers whether this
    /// member may read the event at all. Observed rather than private because
    /// the organiser screen shows one section per form and names each one.
    private(set) var forms: [String: [Form]] = [:]
    private(set) var isLoadingRoster = false
    private(set) var isSubmitting = false
    private(set) var errorMessage: String?

    /// How many check-ins this app has written for an event, used only to tell
    /// a refetch that started earlier that it is now out of date.
    @ObservationIgnored private var writes: [String: Int] = [:]

    /// The events with a roster fetch already in flight, so a poll coming round
    /// again does not start a second one behind it.
    @ObservationIgnored private var inFlight: Set<String> = []

    private static let indicoHost = "event.stsa.tw"
    private static let host = "https://\(indicoHost)"

    func access(for eventID: String) -> Access { access[eventID] ?? .unknown }

    func forms(for eventID: String) -> [Form] { forms[eventID] ?? [] }

    func entries(for eventID: String) -> [Entry] { roster[eventID] ?? [] }

    /// One form's registrants — the list a door actually works from.
    func entries(for eventID: String, formID: Int) -> [Entry] {
        entries(for: eventID).filter { $0.formID == formID }
    }

    /// That list minus the cancelled rows — the people a door is waiting for.
    ///
    /// Withdrawn and rejected registrations stay on Indico's list, because
    /// `~is_deleted` is the only filter its API applies, and out of every count
    /// the app shows, because nobody is expecting them. Indico's own
    /// `active_registration_count` draws the line in the same place, which is
    /// what keeps these numbers comparable with its management page.
    func expected(eventID: String, formID: Int) -> [Entry] {
        entries(for: eventID, formID: formID).filter { !$0.registration.isCancelled }
    }

    // MARK: - The numbers

    /// The counts a screen shows, answered from the roster when there is one and
    /// from the probe's snapshot until then.
    ///
    /// The order matters after a scan: the probe's numbers were taken before the
    /// door opened, while the roster is folded forward by [checkIn] as people
    /// come through. It also means the number can shift slightly when the list
    /// lands and replaces it — see [Form.registrationCount] for why.
    func checkedInCount(eventID: String, formID: Int) -> Int {
        if let loaded = loadedExpected(eventID: eventID, formID: formID) {
            return loaded.filter(\.registration.checkedIn).count
        }
        return form(eventID: eventID, formID: formID)?.checkedInCount ?? 0
    }

    func registeredCount(eventID: String, formID: Int) -> Int {
        if let loaded = loadedExpected(eventID: eventID, formID: formID) { return loaded.count }
        return form(eventID: eventID, formID: formID)?.registrationCount ?? 0
    }

    /// The whole event, for the one-line summary on the event screen.
    func checkedInCount(eventID: String) -> Int {
        if roster[eventID] != nil {
            return entries(for: eventID)
                .filter { !$0.registration.isCancelled && $0.registration.checkedIn }
                .count
        }
        return forms(for: eventID).reduce(0) { $0 + $1.checkedInCount }
    }

    func registeredCount(eventID: String) -> Int {
        if roster[eventID] != nil {
            return entries(for: eventID).filter { !$0.registration.isCancelled }.count
        }
        return forms(for: eventID).reduce(0) { $0 + $1.registrationCount }
    }

    /// The form's expected list, or nil when the roster has not landed — the
    /// difference between "nobody" and "not known yet", which is the whole
    /// question the fallback turns on. An empty roster is an answer.
    private func loadedExpected(eventID: String, formID: Int) -> [Entry]? {
        guard roster[eventID] != nil else { return nil }
        return expected(eventID: eventID, formID: formID)
    }

    private func form(eventID: String, formID: Int) -> Form? {
        forms(for: eventID).first { $0.id == formID }
    }

    /// Asks Indico whether this member manages the event, which is the same
    /// request that fetches the form ids the roster needs.
    func probe(eventID: String, using indico: IndicoAuthManager) async {
        guard access(for: eventID) == .unknown, indico.isLinked else { return }
        access[eventID] = .checking

        guard let url = URL(string: "\(Self.host)/api/checkin/event/\(eventID)/forms/") else {
            access[eventID] = .denied
            return
        }

        do {
            let request = try indico.authorizedRequest(for: url)
            let (data, response) = try await URLSession.shared.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200,
                  let forms = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
            else {
                access[eventID] = .denied
                return
            }

            self.forms[eventID] = forms.compactMap { raw -> Form? in
                guard let id = raw["id"] as? Int else { return nil }
                let title = (raw["title"] as? String ?? "").trimmingCharacters(in: .whitespaces)
                return Form(
                    id: id,
                    title: title,
                    isOpen: raw["is_open"] as? Bool ?? false,
                    registrationCount: raw["registration_count"] as? Int ?? 0,
                    checkedInCount: raw["checked_in_count"] as? Int ?? 0
                )
            }
            access[eventID] = self.forms[eventID]?.isEmpty == false ? .allowed : .denied
        } catch {
            access[eventID] = .denied
        }
    }

    /// Pulls the registrant list once, so a scan is a lookup rather than a
    /// request. The list deliberately carries no answers — Indico excludes them
    /// from the list endpoint — so those are fetched per person on scan.
    func loadRoster(eventID: String, using indico: IndicoAuthManager) async {
        guard roster[eventID] == nil else { return }
        await fetchRoster(eventID: eventID, using: indico)
    }

    /// Asks Indico again, for the door that is not the only one.
    ///
    /// The count on this phone moves when *this* phone records a check-in, and
    /// that is all it knew: a second 幹部 on a second phone, or anyone using
    /// Indico's own app, moved a number this one never saw. So the list is
    /// re-asked when the door screen opens, whenever the roster is pulled down,
    /// and every few seconds while either is on screen — rather than being
    /// fetched once and believed for the rest of the event.
    func refreshRoster(eventID: String, using indico: IndicoAuthManager) async {
        await fetchRoster(eventID: eventID, using: indico)
    }

    /// How often a roster on screen re-asks Indico.
    ///
    /// Five seconds because of what the staleness actually costs: a 幹部 waves
    /// somebody through believing they are the first to admit them, and no
    /// screen anywhere will later disagree. It is one request per registration
    /// form — two for 烤場集合 — and only while a 幹部 is looking at the list or
    /// standing at the door, so the traffic is a handful of phones for the hours
    /// an event is being run, not every member in the app.
    static let refreshInterval: Duration = .seconds(5)

    /// Keeps one event's roster current for as long as the calling task lives.
    ///
    /// A loop the caller owns rather than a timer the store owns: a screen going
    /// away, or the app leaving the foreground, cancels the task and the polling
    /// stops with it. Nothing re-asks Indico about a door nobody is standing at.
    ///
    /// Asks before it waits, because the caller restarts this every time the app
    /// returns to the foreground — and a phone that has been in a pocket for ten
    /// minutes is holding the worst list in the building. Waiting first would
    /// show it for five more seconds to somebody already reading it. The extra
    /// request this costs on the way in is usually not made at all: every screen
    /// here also fetches on appear, and `inFlight` folds the two together.
    func autoRefresh(eventID: String, using indico: IndicoAuthManager) async {
        while !Task.isCancelled {
            await fetchRoster(eventID: eventID, using: indico)
            try? await Task.sleep(for: Self.refreshInterval)
        }
    }

    private func fetchRoster(eventID: String, using indico: IndicoAuthManager) async {
        let eventForms = forms(for: eventID)
        guard !eventForms.isEmpty else { return }

        // One fetch per event at a time. Polling on a venue's wifi will sooner
        // or later come round before the last request landed, and stacking them
        // buys nothing: they all ask the same question, and the slowest would
        // answer it last.
        guard !inFlight.contains(eventID) else { return }
        inFlight.insert(eventID)
        defer { inFlight.remove(eventID) }

        // The flag behind a first-load spinner, and nothing else: once there is
        // a list on screen, a refresh must not take it away and put a spinner
        // there — least of all one arriving every five seconds.
        let isFirstLoad = roster[eventID] == nil
        if isFirstLoad { isLoadingRoster = true }
        defer { if isFirstLoad { isLoadingRoster = false } }

        // Read before the requests go out and compared after they come back: a
        // check-in recorded while this was in flight is newer than anything the
        // response can contain, and letting a stale list land on top of it would
        // take the person back off the screen they were just admitted on.
        let generation = writes[eventID, default: 0]

        var entries: [Entry] = []
        var isComplete = true
        for form in eventForms {
            guard let url = URL(string:
                "\(Self.host)/api/checkin/event/\(eventID)/forms/\(form.id)/registrations/")
            else { isComplete = false; continue }

            do {
                let request = try indico.authorizedRequest(for: url)
                let (data, response) = try await URLSession.shared.data(for: request)
                guard (response as? HTTPURLResponse)?.statusCode == 200 else {
                    isComplete = false
                    continue
                }
                entries += CheckinDecoder.list(data).map { Entry(formID: form.id, registration: $0) }
            } catch {
                isComplete = false
                errorMessage = error.localizedDescription
            }
        }

        // Cancellation is routine now that this is polled — a screen going away
        // or the app backgrounding takes every request in the loop above down
        // with it. Those all landed in `catch` and read as a venue with no wifi,
        // which on a first load would store the empty list they produced.
        guard !Task.isCancelled else { return }

        guard writes[eventID, default: 0] == generation else { return }

        // A refresh that could not read every form must not shorten a list the
        // door is working from — the venue's wifi dropping should cost the
        // count its freshness, not its rows.
        guard isComplete || roster[eventID] == nil else { return }

        roster[eventID] = entries
    }

    /// Finds a scanned member in the list of the form whose door is open.
    ///
    /// Matched on email, which is the only thing MembershipAPI and Indico both
    /// know about a person. Case-insensitive because the two do not agree on it;
    /// someone who registered under a different address than their STSA account
    /// will not be found, and the screen says so rather than implying they never
    /// registered.
    ///
    /// Scoped to one form on purpose. A member card names a person, and one
    /// person can hold a registration in several of an event's forms — matching
    /// across all of them recorded the coach passenger's arrival against the
    /// 報名表, or the other way about, depending on nothing.
    func entry(email: String, eventID: String, formID: Int) -> Entry? {
        Self.match(email: email, in: entries(for: eventID, formID: formID))
    }

    /// The event's *other* forms this address is registered in.
    ///
    /// What the door says instead of "not registered" when the member is in the
    /// event but standing at the wrong desk — the case that reads as a duplicate
    /// from the outside and is nothing of the kind.
    func otherForms(email: String, eventID: String, excluding formID: Int) -> [Form] {
        let needle = Self.normalised(email)
        guard !needle.isEmpty else { return [] }

        let found = Set(
            entries(for: eventID)
                .filter { $0.formID != formID && Self.normalised($0.registration.email) == needle }
                .map(\.formID)
        )
        return forms(for: eventID).filter { found.contains($0.id) }
    }

    /// Picks the registration a scan means, out of however many one form holds
    /// for a single address.
    ///
    /// One form can still hold two. Withdrawing and registering again leaves the
    /// old row behind — Indico allows the second one *because* the first is
    /// withdrawn — and a manager adding somebody by hand is warned about the
    /// clash rather than stopped. So the choice is made deliberately instead of
    /// by list order:
    ///
    /// - an admissible one that is already checked in, so a second scan of the
    ///   same person reads 已經報到過 rather than quietly offering to admit
    ///   their other copy and counting one arrival twice;
    /// - failing that an admissible one, since a withdrawn leftover is not the
    ///   answer to "is this person expected";
    /// - failing that the first, so the screen can explain what it found instead
    ///   of claiming they never registered.
    ///
    /// Pure and static so it can be tested without a roster to fetch. Android's
    /// `CheckinSession.match` is the same rule.
    static func match(email: String, in entries: [Entry]) -> Entry? {
        let needle = normalised(email)
        guard !needle.isEmpty else { return nil }

        let matches = entries.filter { normalised($0.registration.email) == needle }
        return matches.first { $0.registration.isAdmissible && $0.registration.checkedIn }
            ?? matches.first { $0.registration.isAdmissible }
            ?? matches.first
    }

    private static func normalised(_ email: String) -> String {
        email.lowercased().trimmingCharacters(in: .whitespaces)
    }

    /// Fetches one registrant's answers, which the roster does not carry.
    func details(for entry: Entry, eventID: String, using indico: IndicoAuthManager) async -> CheckinRegistration? {
        guard let url = URL(string:
            "\(Self.host)/api/checkin/event/\(eventID)/forms/\(entry.formID)/registrations/\(entry.registration.id)")
        else { return nil }

        do {
            let request = try indico.authorizedRequest(for: url)
            let (data, response) = try await URLSession.shared.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return nil }
            return CheckinDecoder.one(data)
        } catch {
            return nil
        }
    }

    /// Resolves a scanned ticket to its registration.
    ///
    /// Unlike a member card, this needs no roster and no email match, so it
    /// works for someone whose Indico address is not their STSA one.
    ///
    /// Indico looks the ticket up globally — `/api/checkin/ticket/<uuid>` carries
    /// no event — and applies the permission check to whatever event it belongs
    /// to. A staffer who manages two events would otherwise resolve, and check
    /// in, a ticket for the wrong one, so the event is compared here.
    func registration(
        ticket: ScannedCode.Ticket,
        eventID: String,
        using indico: IndicoAuthManager
    ) async -> TicketOutcome {
        guard ticket.host == Self.indicoHost else { return .foreignInstance(ticket.host) }
        guard ticket.accompanyingPersonID == nil else { return .accompanyingPerson }

        let uuid = ticket.checkinSecret.uuidString.lowercased()
        guard let url = URL(string: "\(Self.host)/api/checkin/ticket/\(uuid)") else {
            return .unreadable
        }

        do {
            let request = try indico.authorizedRequest(for: url)
            let (data, response) = try await URLSession.shared.data(for: request)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0

            guard status == 200, let registration = CheckinDecoder.one(data) else {
                return status == 403 ? .notPermitted : .unknownTicket
            }
            guard String(registration.eventID) == eventID else { return .otherEvent }
            return .found(registration)
        } catch {
            return .unreachable(error.localizedDescription)
        }
    }

    /// What a scanned ticket turned out to be.
    enum TicketOutcome: Equatable {
        case found(CheckinRegistration)
        /// A ticket this Indico has never issued.
        case unknownTicket
        /// A real ticket, for an event this door is not.
        case otherEvent
        /// Issued by a different Indico instance entirely.
        case foreignInstance(String)
        /// An accompanying person's ticket, which has no registration.
        case accompanyingPerson
        /// The staffer does not manage the event the ticket belongs to.
        case notPermitted
        case unreadable
        case unreachable(String)
    }

    /// Records attendance — or takes it back — and folds Indico's answer into
    /// the roster so the count and any later scan of the same person are right
    /// without refetching.
    ///
    /// Idempotent: PATCHing `checked_in` that is already true is accepted and
    /// keeps the original `checked_in_dt`, so a double scan costs nothing.
    ///
    /// `checkedIn: false` is the undo, and the same endpoint: Indico clears
    /// `checked_in_dt` with the flag, so a mistake at the door leaves no trace
    /// of an arrival that did not happen. It is allowed on a withdrawn
    /// registration where checking one *in* is not — somebody already marked as
    /// arrived who should not have been is exactly who this is for.
    func checkIn(
        _ registration: CheckinRegistration,
        checkedIn: Bool = true,
        eventID: String,
        using indico: IndicoAuthManager
    ) async -> CheckinResult {
        guard !checkedIn || registration.isAdmissible else { return .notAdmissible }
        guard indico.canRecordCheckin else { return .needsAuthorization }

        guard let url = URL(string:
            "\(Self.host)/api/checkin/event/\(registration.eventID)"
            + "/forms/\(registration.formID)/registrations/\(registration.id)")
        else { return .failed(String(localized: "無法組出報到網址。")) }

        isSubmitting = true
        defer { isSubmitting = false }

        do {
            var request = try indico.authorizedRequest(for: url)
            request.httpMethod = "PATCH"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: ["checked_in": checkedIn])

            let (data, response) = try await URLSession.shared.data(for: request)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0

            // 403 here, after a roster loaded fine, means the grant is too narrow
            // rather than the event being someone else's — reading worked.
            if status == 403 { return .needsAuthorization }
            guard status == 200, let updated = CheckinDecoder.one(data) else {
                return .failed(String(localized: "Indico 回應 HTTP \(status)。"))
            }

            replace(updated, eventID: eventID)
            return .recorded(updated)
        } catch {
            return .failed(error.localizedDescription)
        }
    }

    enum CheckinResult: Equatable {
        /// Indico took the write. Whether that was an arrival or an undo is on
        /// the registration it hands back, not on the case.
        case recorded(CheckinRegistration)
        /// Withdrawn or rejected — not someone to admit.
        case notAdmissible
        /// The Indico authorization is read-only; the staffer must grant
        /// `registrants` before anything can be written.
        case needsAuthorization
        case failed(String)
    }

    private func replace(_ registration: CheckinRegistration, eventID: String) {
        writes[eventID, default: 0] += 1

        guard var entries = roster[eventID],
              let index = entries.firstIndex(where: { $0.registration.id == registration.id })
        else { return }
        entries[index] = Entry(formID: entries[index].formID, registration: registration)
        roster[eventID] = entries
    }

    /// Called when the member signs out — a roster is other people's data.
    func clear() {
        access.removeAll()
        roster.removeAll()
        forms.removeAll()
        writes.removeAll()
        inFlight.removeAll()
        errorMessage = nil
    }
}

File: android/app/src/main/java/tw/stsa/memberapp/feature/checkin/CheckinStore.kt
package tw.stsa.memberapp.feature.checkin

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.delay
import tw.stsa.memberapp.auth.IndicoAuthManager
import tw.stsa.memberapp.model.CheckinRegForm
import tw.stsa.memberapp.model.CheckinRegistration

/**
 * The organiser side of an event: whether this account runs it, and who has
 * turned up.
 *
 * The counterpart of iOS's `CheckinStore`, and the layer *outside* the scanner.
 * [CheckinSession] is still what a worker holds at a door — one event, one form,
 * the list in memory for offline matching. This lives for the whole app, because
 * the event page needs the answer before anyone opens a camera.
 *
 * **Asked, not assumed.** The entry to 報到 used to be gated on the `isOfficer`
 * group claim, which answers a different question: a 幹部 is not necessarily a
 * 幹部 *of this event*, so the button appeared for people whose first tap was a
 * 403. Indico decides instead — `/api/checkin/event/<id>/forms/` is
 * `RHManageEventBase`, so a successful call is the permission, and the group
 * claim goes back to being what `Profile.groups` says it is: not a security
 * boundary, and not used as one.
 *
 * **Why the counts are free.** That same probe answers with each form's
 * `registration_count` and `checked_in_count`, so the row can say "12 / 40
 * 已報到" off one cheap request. iOS pulls the whole roster to work the same
 * number out; there is no reason to here, and the roster is left for the screen
 * that actually lists people.
 */
class CheckinStore {

    enum class Access {
        UNKNOWN,
        CHECKING,

        /** This account manages the event and may read its registrations. */
        ALLOWED,
        DENIED,
    }

    private val access = mutableStateMapOf<String, Access>()
    private val forms = mutableStateMapOf<String, List<CheckinRegForm>>()
    private val rosters = mutableStateMapOf<String, List<CheckinRegistration>>()

    /**
     * How many check-ins this app has written for an event, used only to tell a
     * refetch that started earlier that it is now out of date.
     */
    private val writes = mutableMapOf<String, Int>()

    /**
     * The events with a roster fetch already in flight, so a poll coming round
     * again does not start a second one behind it.
     */
    private val inFlight = mutableSetOf<String>()

    var isLoadingRoster by mutableStateOf(false)
        private set

    /** True while a check-in written from the roster is in flight. */
    var isSubmitting by mutableStateOf(false)
        private set

    fun access(eventId: String): Access = access[eventId] ?: Access.UNKNOWN

    fun forms(eventId: String): List<CheckinRegForm> = forms[eventId].orEmpty()

    fun registrations(eventId: String): List<CheckinRegistration> = rosters[eventId].orEmpty()

    /** One form's registrants — the list a door actually works from. */
    fun registrations(eventId: String, formId: Int): List<CheckinRegistration> =
        registrations(eventId).filter { it.regformId == formId }

    /**
     * How many have arrived, out of how many are coming — for one form.
     *
     * Per form because an event's forms are separate lists: 烤場集合 runs a 報名表
     * and a 遊覽車報名表, and a member on both is one person holding two
     * registrations rather than a duplicate of themselves. Added together they
     * made a denominator no door could reach.
     *
     * Read from the roster once there is one and from the probe's own totals
     * until then. That order matters after a scan: the probe's numbers are a
     * snapshot from before the door opened, while the roster is folded forward
     * by [update] as people are checked in.
     *
     * Withdrawn and rejected registrations are in neither half once the roster
     * lands — nobody is waiting for them at a door. Indico's own
     * `registration_count` is `existing_registrations_count`, which counts them
     * and counts accompanying persons as seats, so the number can shift a little
     * when the list arrives and replaces it.
     */
    fun checkedInCount(eventId: String, formId: Int): Int =
        expected(eventId, formId)?.count { it.checkedIn }
            ?: forms(eventId).firstOrNull { it.id == formId }?.checkedInCount
            ?: 0

    fun registeredCount(eventId: String, formId: Int): Int =
        expected(eventId, formId)?.size
            ?: forms(eventId).firstOrNull { it.id == formId }?.registrationCount
            ?: 0

    /** The whole event, for the one-line summary on the event page. */
    fun checkedInCount(eventId: String): Int =
        expected(eventId)?.count { it.checkedIn } ?: forms(eventId).sumOf { it.checkedInCount }

    fun registeredCount(eventId: String): Int =
        expected(eventId)?.size ?: forms(eventId).sumOf { it.registrationCount }

    private fun expected(eventId: String): List<CheckinRegistration>? =
        rosters[eventId]?.filter { !it.isCancelled }

    private fun expected(eventId: String, formId: Int): List<CheckinRegistration>? =
        rosters[eventId]?.filter { it.regformId == formId && !it.isCancelled }

    /**
     * Asks Indico whether this account runs the event. Asked once per event per
     * launch: the answer does not change while someone is looking at a screen,
     * and this is on the path of every event anyone opens.
     */
    suspend fun probe(eventId: String, indico: IndicoAuthManager) {
        if (access(eventId) != Access.UNKNOWN || !indico.isLinked) return
        // The check-in API is addressed by numeric id; an event whose id is not
        // one is not an event this API knows.
        val numericId = eventId.toIntOrNull() ?: run {
            access[eventId] = Access.DENIED
            return
        }

        access[eventId] = Access.CHECKING
        try {
            val found = IndicoCheckinClient(indico).regforms(numericId)
            forms[eventId] = found
            access[eventId] = if (found.isNotEmpty()) Access.ALLOWED else Access.DENIED
        } catch (error: CheckinError) {
            // 403 is the ordinary answer for a member who does not staff this
            // event, and it is not worth a message: the row simply is not there.
            access[eventId] = Access.DENIED
        }
    }

    /** The list itself, which only the organiser screen needs. */
    suspend fun loadRoster(eventId: String, indico: IndicoAuthManager) {
        if (rosters.containsKey(eventId)) return
        fetchRoster(eventId, indico)
    }

    /**
     * Asks Indico again, for the door that is not the only one.
     *
     * The count on this phone moved when *this* phone recorded a check-in, and
     * that was all it knew: a second 幹部 on a second phone, or anyone using
     * Indico's own app, moved a number this one never saw. The list was fetched
     * once per launch and believed for the rest of the event.
     */
    suspend fun refreshRoster(eventId: String, indico: IndicoAuthManager) {
        fetchRoster(eventId, indico)
    }

    /**
     * Keeps one event's roster current for as long as the calling coroutine
     * lives.
     *
     * A loop the caller owns rather than a timer the store owns: a screen going
     * away, or the app leaving the foreground, cancels the coroutine and the
     * polling stops with it. Nothing re-asks Indico about a door nobody is
     * standing at.
     *
     * Asks before it waits, because the caller restarts this every time the app
     * returns to the foreground — and a phone that has been in a pocket for ten
     * minutes is holding the worst list in the building. Waiting first would
     * show it for five more seconds to somebody already reading it. The extra
     * request this costs on the way in is usually not made at all: every screen
     * here also fetches on appear, and [inFlight] folds the two together.
     */
    suspend fun autoRefresh(eventId: String, indico: IndicoAuthManager) {
        while (true) {
            fetchRoster(eventId, indico)
            delay(REFRESH_INTERVAL_MS)
        }
    }

    private suspend fun fetchRoster(eventId: String, indico: IndicoAuthManager) {
        val numericId = eventId.toIntOrNull() ?: return
        val known = forms(eventId).ifEmpty { return }

        // One fetch per event at a time. Polling on a venue's wifi will sooner
        // or later come round before the last request landed, and stacking them
        // buys nothing: they all ask the same question, and the slowest would
        // answer it last.
        if (!inFlight.add(eventId)) return

        // The flag behind a first-load spinner, and nothing else: once there is
        // a list on screen, a refresh must not take it away and put a spinner
        // there — least of all one arriving every five seconds.
        val isFirstLoad = !rosters.containsKey(eventId)
        if (isFirstLoad) isLoadingRoster = true

        // Read before the requests go out and compared after they come back: a
        // check-in recorded while this was in flight is newer than anything the
        // response can contain, and letting a stale list land on top of it would
        // take the person back off the screen they were just admitted on.
        val generation = writes[eventId] ?: 0

        try {
            val client = IndicoCheckinClient(indico)
            var isComplete = true
            val entries = known.flatMap { form ->
                try {
                    client.registrations(numericId, form.id)
                } catch (cancelled: CancellationException) {
                    // Routine now that this is polled: a screen going away or
                    // the app backgrounding cancels mid-request. Swallowed as an
                    // ordinary failure it would read as a venue with no wifi,
                    // and on a first load store the empty list that produced.
                    throw cancelled
                } catch (error: Exception) {
                    isComplete = false
                    emptyList()
                }
            }

            if ((writes[eventId] ?: 0) != generation) return

            // A refresh that could not read every form must not shorten a list
            // the door is working from — the venue's wifi dropping should cost
            // the count its freshness, not its rows.
            if (!isComplete && !isFirstLoad) return

            rosters[eventId] = entries
        } finally {
            inFlight.remove(eventId)
            if (isFirstLoad) isLoadingRoster = false
        }
    }

    /**
     * Folds a scan back into the roster, so the count on the way out of the
     * scanner is the count the door just produced rather than the one it started
     * with. [CheckinSession] does the same to its own copy; this is the shared
     * one behind it.
     */
    fun update(registration: CheckinRegistration) {
        val eventId = registration.eventId.toString()
        writes[eventId] = (writes[eventId] ?: 0) + 1
        val current = rosters[eventId] ?: return
        rosters[eventId] = current.map { if (it.id == registration.id) registration else it }
    }

    /**
     * Records attendance, or takes it back, from the roster rather than a door.
     *
     * [CheckinSession] does this for a scan; this is the same write for the list
     * on 幹部功能, where a name is tapped instead. Deliberately the same flag and
     * the same endpoint — a check-in made from the list is not a second kind of
     * check-in, and Indico cannot tell them apart.
     *
     * A withdrawn registration can be *un*-checked-in but not checked in: the
     * one thing this is for is correcting somebody who should never have been
     * marked as arrived.
     */
    suspend fun setCheckedIn(
        registration: CheckinRegistration,
        checkedIn: Boolean,
        indico: IndicoAuthManager,
    ): CheckinRegistration {
        if (!indico.canRecordCheckin) throw CheckinError.NeedsAuthorization
        if (checkedIn && !registration.isAdmissible) {
            throw CheckinError.NotAdmissible(registration.fullName)
        }
        isSubmitting = true
        try {
            val updated = IndicoCheckinClient(indico).checkIn(registration, checkedIn)
            update(updated)
            return updated
        } finally {
            isSubmitting = false
        }
    }

    /**
     * Clears everything. Called from an explicit sign-out, alongside the ticket
     * store — a roster names people, and it should not outlive the account that
     * was allowed to read it.
     */
    fun clear() {
        access.clear()
        forms.clear()
        rosters.clear()
        writes.clear()
        inFlight.clear()
    }

    companion object {
        /**
         * How often a roster on screen re-asks Indico.
         *
         * Five seconds because of what the staleness actually costs: a 幹部 waves
         * somebody through believing they are the first to admit them, and no
         * screen anywhere will later disagree. It is one request per
         * registration form — two for 烤場集合 — and only while a 幹部 is looking
         * at the list or standing at the door, so the traffic is a handful of
         * phones for the hours an event is being run, not every member in the
         * app. iOS's `CheckinStore.refreshInterval` is the same number.
         */
        const val REFRESH_INTERVAL_MS = 5_000L
    }
}
